### PR TITLE
CI: Implement mypy pre-commit check

### PR DESCRIPTION
**Pull request by ghost**
* Add mypy check in pre-commit
* Fix typing in various places
* Split consumer into `BaseConsumer`/`Consumer`/`RowConsumer` to support accepting different topic type (`str` and `Topic`) in `subscribe`

### CODE REVIEW

File: quixstreams/app.py [hunk 1]
```diff
@@ -1009,8 +1023,8 @@ def copy(self, **kwargs) -> Self:
         return copy
 
     @property
-    def flush_timeout(self) -> float:
-        return (
+    def flush_timeout(self) -> int:
+        return int(
```
Comment (ghost) [inline quixstreams/app.py]: I'm unsure if everything should be int or float. Currently most timeout expect an int

File: requirements-mypy.txt [hunk 2]
```diff
@@ -0,0 +1,5 @@
+mypy==1.13.0
```
Comment (gwaramadze) [inline requirements-mypy.txt:1]: Shouldn't mypy be installed by pre-commit?

Comment (ghost) [reply requirements-mypy.txt:1]: We need pre-commit to use the system environment for mypy. Mypy needs to run on a environment with all the dependencies installed. There is no good way to have a pre-commit environment that does that, using the system environment is easier (system mean the current active environment outside of pre-commit not the system base python install)

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -24,7 +24,13 @@ jobs:
...
         uses: actions/setup-python@v4
         with:
-          python-version: 3.9
+          python-version: 3.12
+      - name: Update pip
+        run: |
+          python -m pip install -U pip
+      - name: Install requirements
+        run: |
+          python -m pip install -U -r requirements.txt -r tests/requirements.txt -r requirements-mypy.txt
```
Comment (gwaramadze) [inline .github/workflows/ci.yml]: I think we can zip both steps together

```suggestion
      - name: Install requirements
        run: |
          python -m pip install -U pip
          python -m pip install -U -r requirements.txt -r tests/requirements.txt -r requirements-mypy.txt
```

Comment (gwaramadze) [reply .github/workflows/ci.yml]: If you choose to do it, you can do the same for `test` job.

File: requirements-mypy.txt (hunk 2 above)
Comment (gwaramadze) [reply requirements-mypy.txt:1]: Okay, I see that you use `language: system` so we need to install it outside of pre-commit env

File: .pre-commit-config.yaml [hunk 4]
```diff
@@ -11,3 +11,10 @@ repos:
         entry: python conda/requirements.py
         language: python
         files: ^(requirements\.txt|pyproject\.toml)$
+  - repo: https://github.com/pre-commit/mirrors-mypy
+    rev: v1.13.0
+    hooks:
+    -   id: mypy
+        args: []
```
Comment (gwaramadze) [inline .pre-commit-config.yaml:18]: I think this may be removed
```suggestion
```

File: pyproject.toml [hunk 5]
```diff
@@ -96,3 +96,39 @@ log_cli_level = "INFO"
...
+[[tool.mypy.overrides]]
+module = "quixstreams.sources.community.*"
+ignore_errors = true
+
+[[tool.mypy.overrides]]
+module = "quixstreams.state.*"
+ignore_errors = true
+
+[[tool.mypy.overrides]]
+module = "quixstreams.rowproducer.*"
```
Comment (gwaramadze) [inline pyproject.toml]: This should do

```suggestion
module = [
    "quixstreams.core.*",
    "quixstreams.dataframe.*",
    "quixstreams.models.*",
    "quixstreams.platforms.*",
    "quixstreams.sinks.community.*",
    "quixstreams.sources.community.*",
    "quixstreams.state.*",
    "quixstreams.rowproducer.*"
]
```

Comment (gwaramadze) [reply pyproject.toml]: So many ignored 🙈
Too much work to fix those?

File: .pre-commit-config.yaml (hunk 4 above)
Comment (ghost) [reply .pre-commit-config.yaml:18]: It's needed to override the one [set in the recipe repo](https://github.com/pre-commit/mirrors-mypy/blob/main/.pre-commit-hooks.yaml#L7C5-L7C64)

I prefer to configure mypy using pyproject.toml so we can easily run it manually

File: pyproject.toml (hunk 5 above)
Comment (ghost) [reply pyproject.toml]: Too much work for this PR. The goal is to fix them at one point.

**Review by gwaramadze — DISMISSED** [review]

**Review by gwaramadze — APPROVED** [review]

**Pull request merged**